### PR TITLE
Separate address node of BCDCHK and fix data type

### PR DESCRIPTION
**Pull request by NigelYiboYu**
BCDCHK is a node that can GC-and-Return. As a result of this,
LocalCSE does not common BCDCHK's address node with any address
node after the BCDCHK.

When the BCDCHK produces a PackedDecimal type, it'll be followed by a
pdstorei node, which has the same address node as the BCDCHK node.
If this BCDCHK and it's following pdstorei share the same
address node, then no address node after the pdstorei could be commoned
with the pdstorei's address node because BCDCHK prevents this from
happening. Failing to common such address nodes can result in unnecessary
load-after-store instruction sequence (e.g. VLRL after VSTRL). This is
redundant and inefficient.

Separating BCDCHK's and pdstorei's address nodes makes the aforementioned
address node commoning possible and opens up opportunities for packed
decimal copy propagations.

Also, fix the symbol reference data type for the pdstorei node of
I2PD and PD-shift-left. The stored type should be TR::PackedDecimal
instead of TR::Int8.

Signed-off-by: Nigel Yu <yunigel@ca.ibm.com>

### CODE REVIEW

**Comment by NigelYiboYu** [comment]
Marking this WIP until my tests are completed.

File: runtime/compiler/trj9/optimizer/DataAccessAccelerator.cpp [hunk 1]
```diff
@@ -1218,9 +1218,54 @@ bool TR_DataAccessAccelerator::generateI2PD(TR::TreeTop* treeTop, TR::Node* call
...
+       * See BCDCHKEvaluatorImpl() for BCDCHK tree structure and intended use of its children.
+       *
+       * 'outOfLineCopyBackAddr' and 'storeAddressNode' correspond to AddrNode1 and AddrNode2, respectively. They
+       * are created as separate nodes so that LocalCSE is able to common up AddrNode2 and AddrNode3. If
+       * AddrNode1 and AddrNode2 were the same node, the LocalCSE would not consider AddrNode1 an alternative replacement
+       * of AddrNode3 because the BCDCHK's symbol canGCAndReturn().
+       *
+       * With AddrNode2 and AddrNode3 commoned up, the LocalCSE is able to copy propagate pdshlOverflow to the the pd2zd
+       * tree and replace its pdloadi.
+       */
```
Comment (fjeremic) [inline runtime/compiler/trj9/optimizer/DataAccessAccelerator.cpp]: Excellent documentation! Future readers, take note 😃

File: runtime/compiler/trj9/z/codegen/J9TreeEvaluator.cpp [hunk 2]
```diff
@@ -15809,12 +15809,34 @@ J9::Z::TreeEvaluator::BCDCHKEvaluator(TR::Node * node, TR::CodeGenerator * cg)
                break;
                }
 
-            default: TR_ASSERT_FATAL(0, "BCDCHKEvaluator: Could not find recognized method for variable precision DAA operation.\n");
+            default:
+               {
+               /**
+                *
```
Comment (fjeremic) [inline runtime/compiler/trj9/z/codegen/J9TreeEvaluator.cpp]: Extra line here. Can we remove?

File: runtime/compiler/trj9/z/codegen/J9TreeEvaluator.cpp [hunk 3]
```diff
@@ -15809,12 +15809,34 @@ J9::Z::TreeEvaluator::BCDCHKEvaluator(TR::Node * node, TR::CodeGenerator * cg)
...
          }
 
-      default: TR_ASSERT_FATAL(0, "Operation Code not supported by BCDCHKEvaluator.\n");
+      default:
+         {
+         /**
+          * Unrecognized opCodes under BCDCHK should come from optimizations such as local CSE and tree simplifications.
+          * They should be commoned nodes that's evaluated previously. Skip these nodes.
+          */
+         TR_ASSERT_FATAL(resultReg != NULL, "BCDCHKEvaluator: BCDCHK has an unevaluated non-PD node %p (non-PD op code %s) \n",
```
Comment (fjeremic) [inline runtime/compiler/trj9/z/codegen/J9TreeEvaluator.cpp]: Shouldn't it be the simplifier's job to remove the `BCDCHK` as well if it folds the operation underneath? This seems like special casing a situation that should never occur. The `BCDCHK` IL has a tight specification that it's first child is an operation that may raise a hardware exception. If a `BCDCHK` reaches codegen and this is not the case then the tree verifier (currently does not run in production builds) should have asserted that this is invalid IL.

File: runtime/compiler/trj9/z/codegen/J9TreeEvaluator.cpp (hunk 2 above)
Comment (NigelYiboYu) [reply runtime/compiler/trj9/z/codegen/J9TreeEvaluator.cpp]: Removed the extra line.

File: runtime/compiler/trj9/z/codegen/J9TreeEvaluator.cpp (hunk 3 above)
Comment (NigelYiboYu) [reply runtime/compiler/trj9/z/codegen/J9TreeEvaluator.cpp]: In theory, yes we can make the simplifier remove such `BCDCHK` nodes. There's one catch: the call parameter nodes under the `BCDCHK` has to be anchored somewhere and get evaluated. This is especially important for the single-reference-unevaluated call parameter nodes.

Shall we open an issue for the simplifier?

Comment (fjeremic) [reply runtime/compiler/trj9/z/codegen/J9TreeEvaluator.cpp]: > Shall we open an issue for the simplifier?

@NigelYiboYu yes I think so. Please reference this assert in the item you open. This assertion should happen at the tree verification level, i.e. check first child of `BCDCHK` and make sure it is a BCD operation.

**Comment by fjeremic** [comment]
Jenkins test all

**Comment by fjeremic** [comment]
AIX build has an infrastructure failure and the Travis build was blocked by #1711 at the time of the latest commit. Everything else looks good for a merge.

**Pull request merged**